### PR TITLE
feat: don't allocate in UDP recv path

### PR DESCRIPTION
**Pull request by mxinden**
This is the final stacked pull request combining the following commits:

**feat(common): make Datagram generic over payload type**

Previously the payload type of `Datagram` was `Vec<u8>`, thus the `Datagram` always owned the UDP datagram payload.

This change enables `Datagram` to represent both (a) an owned payload allocation (i.e. `Vec<u8>`), but also represent (b) a view into an existing payload (e.g. a long lived receive buffer via `&[u8]`).

The default payload type stays `Vec<u8>`, thus not breaking existing usage of `Datagram`.

**feat(transport): accept borrowed instead of owned input datagram**

Previously `process_input` (and the like) took a `Datagram<Vec<u8>>` as input. In other words, it required allocating each UDP datagram payload in a dedicated `Vec<u8>` before passing it to `process_input`.

With this patch, `process_input` accepts a `Datagram<&[u8]>`. In other words, it accepts a `Datagram` with a borrowed view into an existing buffer (`&[u8]`), e.g. a long lived receive buffer.

**feat(udp): return borrowed Datagram on receive**

Previously `recv_inner` would return `Datagram<Vec<u8>>`. In other words, it would allocate a new `Vec<u8>` for each UDP datagram payload.

Now `recv_inner` reads into a provided buffer and returns `Datagram<&[u8]>`, i.e. it returns a view into the provided buffer without allocating.

**feat(bin): don't allocate in UDP recv path**

Pass a long lived receive buffer to `neqo_udp::recv_inner`, receiving an iterator of `Datagram<&[u8]>`s pointing into that buffer, thus no longer allocating in UDP receive path.

---

Extracted out of https://github.com/mozilla/neqo/pull/2093.

Part of https://github.com/mozilla/neqo/issues/1693.

[Pull request stack](https://github.com/marketplace/stacked-pull-requests):

1.  (this one) https://github.com/mozilla/neqo/pull/2184
2. https://github.com/mozilla/neqo/pull/2187
3. https://github.com/mozilla/neqo/pull/2188
4. https://github.com/mozilla/neqo/pull/2189

### CODE REVIEW

**Comment by mxinden** [comment]
This is the final pull request in the stack. All commits, but the first (https://github.com/mozilla/neqo/pull/2184/commits/08a0003b2a5deb3bdf4c787d874d27d8f9e79275), have been reviewed and merged through the 3 previous stacked pull requests.

@martinthomson and @larseggert can you do one more review?

File: neqo-common/src/datagram.rs
Comment (martinthomson) [inline neqo-common/src/datagram.rs:50]: It's harmless if not, though I think it might make some of our tests easier.

File: neqo-udp/src/lib.rs
Comment (martinthomson) [inline neqo-udp/src/lib.rs:107]: Code coverage suggests that this might be something worth testing.

**Review by martinthomson — APPROVED** [review]
LGTM.

File: neqo-common/src/datagram.rs
Comment (mxinden) [reply neqo-common/src/datagram.rs]: Removed `TODO` and moved behing `#[cfg(test)]`. In case we ever need it in non-tests, we can remove the `cfg`.

File: neqo-udp/src/lib.rs
Comment (mxinden) [reply neqo-udp/src/lib.rs:107]: :+1: Added a basic test.

**Comment by larseggert** [comment]
@mxinden is this good to go? Nothing more pending?

**Comment by mxinden** [comment]
Nothing else pending from my end @larseggert. Please merge in case you have no further comments.

I will propose a new release once merged. Excited to get this into Firefox.

**Comment by mxinden** [comment]
Thank you for the help here!

For early CPU profiles of this patch in Firefox see https://github.com/mozilla/neqo/pull/2198#issue-2611368183

**Pull request merged**
